### PR TITLE
OnTouch support when npc is walking

### PR DESCRIPTION
**Pull request by Atemo**
* **Addressed Issue(s)**: https://github.com/rathena/rathena/issues/502

* **Server Mode**: -

* **Description of Pull Request**: 

Added support to trigger 'OnTouch'/'OnTouch_' npc label when a npc is moving through npcwalkto script command.

### CODE REVIEW

File: src/map/npc.cpp [hunk 1]
```diff
@@ -1122,14 +1096,71 @@ int npc_touchnext_areanpc(struct map_session_data* sd, bool leavemap)
...
+		pc_setpos(sd, nd->u.warp.mapindex, nd->u.warp.x, nd->u.warp.y, CLR_OUTSIGHT);
+		return 2;
+	case NPCTYPE_SCRIPT:
+		// warp type sorted first, no need to check if they override any other OnTouch areas.
+
+		if (npc_ontouch_event(sd, nd) > 0 && npc_ontouch2_event(sd, nd) > 0) { // failed to run OnTouch event, so just click the npc
+			auto it = std::find(sd->areanpc.begin(), sd->areanpc.end(), nd->bl.id);
+
+			if (it == sd->areanpc.end())
+				sd->areanpc.push_back(nd->bl.id);
```
Comment (aleos89) [inline src/map/npc.cpp:1148]:
```suggestion
			if (!util::vector_exists(sd->areanpc, nd->bl.id))
				sd->areanpc.push_back(nd->bl.id);
```
Don't forget to include the namespace after includes:
`using namespace rathena;`

File: src/map/unit.cpp [hunk 2]
```diff
@@ -316,6 +316,43 @@ TIMER_FUNC(unit_step_timer){
...
+	case BL_PC:
+		TBL_PC *sd = (TBL_PC*)bl;
+
+		if (sd == nullptr)
+			return 1;
+		if (sd->state.block_action & PCBLOCK_NPCCLICK)
+			return 1;
+
+		// Remove NPCs that are no longer within the OnTouch area
+		for (int i = 0; i < sd->areanpc.size(); i++) {
```
Comment (aleos89) [inline src/map/unit.cpp:336]:
```suggestion
		for (size_t i = 0; i < sd->areanpc.size(); i++) {
```

File: src/map/unit.cpp [hunk 3]
```diff
@@ -316,6 +316,43 @@ TIMER_FUNC(unit_step_timer){
...
+			struct npc_data *nd = map_id2nd(sd->areanpc[i]);
+
+			if (!nd || nd->subtype != NPCTYPE_SCRIPT ||
+				!(nd->bl.m == bl->m && bl->x >= nd->bl.x - nd->u.scr.xs && bl->x <= nd->bl.x + nd->u.scr.xs && bl->y >= nd->bl.y - nd->u.scr.ys && bl->y <= nd->bl.y + nd->u.scr.ys))
+				sd->areanpc.erase(sd->areanpc.begin() + i);
```
Comment (aleos89) [inline src/map/unit.cpp:341]:
```suggestion
				util::erase_at(sd->areanpc, i);
```

File: src/map/unit.cpp [hunk 4]
```diff
@@ -493,6 +532,24 @@ static TIMER_FUNC(unit_walktoxy_timer){
...
+			case NPCTYPE_SCRIPT:
+				xs = nd->u.scr.xs;
+				ys = nd->u.scr.ys;
+				break;
+			case NPCTYPE_WARP:
+				xs = nd->u.warp.xs;
+				ys = nd->u.warp.ys;
+				break;
+			}
+			if (xs > -1 || ys > -1)
```
Comment (aleos89) [inline src/map/unit.cpp:550]: This doesn't require both values to be > -1? To assume it's on a valid cell.

Comment (Atemo) [reply src/map/unit.cpp:550]: Hmmm you're right I followed a mistake in map.cpp

File: src/map/npc.cpp [hunk 5]
```diff
@@ -1122,22 +1089,75 @@ int npc_touchnext_areanpc(struct map_session_data* sd, bool leavemap)
...
 /*==========================================
  * Exec OnTouch for player if in range of area event
  *------------------------------------------*/
-int npc_touch_areanpc(struct map_session_data* sd, int16 m, int16 x, int16 y)
+int npc_touch_area_allnpc(struct map_session_data* sd, int16 m, int16 x, int16 y)
 {
-	int xs, ys, i, f = 1;
-
 	nullpo_retr(1, sd);
+	int i, f = 1;
```
Comment (aleos89) [inline src/map/npc.cpp:1153]: Drop `i` here.

File: src/map/npc.cpp [hunk 6]
```diff
@@ -1122,22 +1089,75 @@ int npc_touchnext_areanpc(struct map_session_data* sd, bool leavemap)
...
 
 	// Remove NPCs that are no longer within the OnTouch area
 	for (i = 0; i < sd->areanpc.size(); i++) {
```
Comment (aleos89) [inline src/map/npc.cpp:1156]:
```suggestion
	for (size_t i = 0; i < sd->areanpc.size(); i++) {
```

File: src/map/npc.cpp [hunk 7]
```diff
@@ -1146,65 +1166,21 @@ int npc_touch_areanpc(struct map_session_data* sd, int16 m, int16 x, int16 y)
 	struct map_data *mapdata = map_getmapdata(m);
 
 	for (i = 0; i < mapdata->npc_num_area; i++) {
```
Comment (aleos89) [inline src/map/npc.cpp:1168]:
```suggestion
	for (int i = 0; i < mapdata->npc_num_area; i++) {
```

**Pull request merged**